### PR TITLE
Incorporate uniprot id into gene info

### PR DESCRIPTION
**Pull request by beatrizsaldana**
## [AG-1392](https://sagebionetworks.jira.com/browse/AG-1392) - Incorporate uniprot ID into gene_info
Adding `uniprotkb_accessions` column to the `gene_info` dataset using the uniprot to ensembl mapping file ([syn54113663](https://www.synapse.org/Synapse:syn54113663)).

### Changes
1. Updated `config.yaml` and `test_config.yaml` to include the uniprot mapping dataset
2. Updated existing `gene_info` tests to include a `uniprotkb_accessions` column
3. Updated `gene_info` output assets to include a `uniprotkb_accessions` column
4. Added `ensg_to_uniprot_mapping_good.tsv` to the `gene_info` testing assets
5. Added gx validation for the `uniprotkb_accessions` column

[AG-1392]: https://sagebionetworks.jira.com/browse/AG-1392?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -169,6 +169,9 @@ datasets:
         - name: tep_adi_info
           id: syn51942280.3
           format: csv
+        - name: ensg_to_uniprot_mapping
+          id: syn54113663
```
Comment (BWMac) [inline config.yaml]: This probably needs a version number i.e. `syn54113663.3`

File: config.yaml [hunk 2]
```diff
@@ -199,6 +204,7 @@ datasets:
         - syn13363443.11
         - *genes_biodomains_provenance
         - syn51942280.3
+        - syn54113663
```
Comment (BWMac) [inline config.yaml]: Same comment as above

File: test_config.yaml [hunk 3]
```diff
@@ -169,6 +169,9 @@ datasets:
         - name: tep_adi_info
           id: syn51942280.3
           format: csv
+        - name: ensg_to_uniprot_mapping
+          id: syn54113663
```
Comment (BWMac) [inline test_config.yaml]: Same version comment

File: test_config.yaml [hunk 4]
```diff
@@ -199,6 +204,7 @@ datasets:
         - syn13363443.11
         - *genes_biodomains_provenance
         - syn51942280.3
+        - syn54113663
```
Comment (BWMac) [inline test_config.yaml]: Same version comment

**Review by BWMac — COMMENTED** [review]
Is the commented out regex GX expectation still meant to be there? - If you wanted to check if the list members matched this regex you could do that by treating the field as nested and using a JSON schema validation expectation.

File: config.yaml (hunk 1 above)
Comment (JessterB) [reply config.yaml]: Yes, let's use v3 here, and in the other three config values

**Review by JessterB — APPROVED** [review]
Pre-approving this, but please add the file version number in the four places Brad flagged. Otherwise we will default to whatever the latest version of the file is, and that can result in unexpected data surprises and murky provenance...

File: tests/test_assets/gene_info/input/ensg_to_uniprot_mapping_good.tsv [hunk 5]
```diff
@@ -0,0 +1,18462 @@
+uniprotkb_accessions	ensembl_gene_id
```
Comment (jaclynbeck-sage) [inline tests/test_assets/gene_info/input/ensg_to_uniprot_mapping_good.tsv:1]: Is there any way you can shorten this file to just a few accession/ensembl pairs? Otherwise it may be hard to add to / edit this test case later.

Comment (beatrizsaldana) [reply tests/test_assets/gene_info/input/ensg_to_uniprot_mapping_good.tsv:1]: Can do!

File: src/agoradatatools/etl/transform/gene_info.py [hunk 6]
```diff
@@ -251,4 +289,7 @@ def transform_gene_info(
     # Make sure there are no N/A Ensembl IDs
     gene_info = gene_info.dropna(subset=["ensembl_gene_id"])
 
+    # Add uniprot ID
+    gene_info = add_uniprot_id_to_gene_info(gene_info, uniprot)
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/gene_info.py]: I would move this to before the .dropna() statement.

File: tests/transform/test_gene_info.py [hunk 7]
```diff
@@ -301,3 +328,63 @@ def test_transform_gene_info_should_fail(
...
+        )
+        uniprot_df = pd.read_csv(
+            os.path.join(
+                self.data_files_path, "input", unpiprot_input_files["uniprot_good"]
+            )
+        )
+        output_df = gene_info.add_uniprot_id_to_gene_info(gene_info_df, uniprot_df)
+
+        # Read in expected dataframe and convert uniprotkb_accessions to list
+        expected_df = pd.read_csv(
```
Comment (jaclynbeck-sage) [inline tests/transform/test_gene_info.py]: Matter of opinion (so, optional): It might be cleaner to save the expected output as a json file so no conversion code is needed.

File: tests/transform/test_gene_info.py [hunk 8]
```diff
@@ -301,3 +328,63 @@ def test_transform_gene_info_should_fail(
                 adjusted_p_value_threshold=param_set["adjusted_p_value_threshold"],
                 protein_level_threshold=param_set["protein_level_threshold"],
             )
+
+    @pytest.mark.parametrize(
+        "unpiprot_input_files, expected_output_file",
+        pass_uniprot_test_data,
+        ids=pass_uniprot_test_ids,
+    )
+    def test_add_uniprot_id_to_gene_info_should_pass(
```
Comment (jaclynbeck-sage) [inline tests/transform/test_gene_info.py]: I'm a little stuck on whether this belongs in this file as an integration test or whether this is considered a unit test -- let me think on it.

File: src/agoradatatools/etl/transform/gene_info.py (hunk 6 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/gene_info.py]: Actually, now that I look at the whole function -- I would prefer that the uniprot data set gets added in the same way as all the other data sets, for consistency. So, all of the other input data sets have whatever necessary processing done to them pre-line-173. Then line 173 begins a loop over all data sets to do outer joins on `ensembl_gene_id`. So your `add_uniprot_id_to_gene_info()` function should actually be a function that creates `collapsed_uniprot` and returns it, and then that df gets added to the loop at line 173. Does that make sense?
This also simplifies testing that function vs testing the whole transform, in my opinion.

Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/gene_info.py]: This makes sense! Thank you for catching it! In an effort to cause the lest amount of code changes (and therefore testing) I missed this opportunity for better integration.

Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/gene_info.py]: Also just FYI, everything is an outer join instead of a left join because while we try to keep `gene_metadata` up to date to have all genes in all data sets, it is possible that adding a new data set or re-running pre-processing on one of the datasets we process could introduce genes that aren't in `gene_metadata` yet, and we wanted those genes to get added.

Although, now I'm wondering if we _should_ do left joins instead and throw a warning or error to update `gene_info`. Not your problem to fix right now but, worth discussing some time.

File: src/agoradatatools/etl/transform/gene_info.py [hunk 9]
```diff
@@ -5,6 +5,43 @@
...
+    gene_info = pd.merge(
+        left=gene_info, right=collapsed_uniprot, on="ensembl_gene_id", how="left"
+    )
+
+    return gene_info
+
+
 def transform_gene_info(
     datasets: dict, adjusted_p_value_threshold: float, protein_level_threshold: float
 ) -> pd.DataFrame:
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/gene_info.py]: While you're editing this file, can you fix a typo I made in the function's main comment? It says "Each dataset will be left_joined onto gene_info", and it should be outer_joined.

File: tests/transform/test_gene_info.py (hunk 7 above)
Comment (beatrizsaldana) [reply tests/transform/test_gene_info.py]: So true, I like your suggestion better than what I implemented, thank you!

File: tests/transform/test_gene_info.py (hunk 8 above)
Comment (beatrizsaldana) [reply tests/transform/test_gene_info.py]: Now that it is a `collapse_uniprot_df()` function, it feels more like a unit test. I'll move it out of here.

File: tests/test_assets/gene_info/input/ensg_to_uniprot_mapping_good.tsv (hunk 5 above)
Comment (jaclynbeck-sage) [reply tests/test_assets/gene_info/input/ensg_to_uniprot_mapping_good.tsv:1]: Thanks. Are there any repeat Ensembl IDs in this file? If not, I think it's worth adding at least one.

File: tests/transform/test_gene_info.py [hunk 10]
```diff
@@ -70,6 +70,11 @@
                        or is_tep values are assumed to mean "False". These two fields must have boolean values if the
```
Comment (jaclynbeck-sage) [inline tests/transform/test_gene_info.py:70]: Make sure to update the comment to say "This transform requires **13** different input datasets" now that we've added one!

File: gx_suite_definitions/gene_info.ipynb [hunk 11]
```diff
@@ -343,6 +343,20 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# uniprotkb_accessions\n",
+    "validator.expect_column_values_to_be_of_type(\"uniprotkb_accessions\", \"list\")\n",
+    "validator.expect_column_values_to_have_list_members_of_type(column=\"uniprotkb_accessions\", member_type=\"str\")\n",
+    "\n",
+    "# Regex from https://www.uniprot.org/help/accession_numbers\n",
+    "# validator.expect_column_values_to_match_regex(\"uniprotkb_accession\", \"[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}\")"
```
Comment (jaclynbeck-sage) [inline gx_suite_definitions/gene_info.ipynb]: Is this commented out for a reason? I feel like I remember us having some issues with the regex for UniProt accessions in a different PR but I can't remember if we figured it out or gave up...

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
Needs a little bit of moving code around but it's almost there!

File: gx_suite_definitions/gene_info.ipynb (hunk 11 above)
Comment (beatrizsaldana) [reply gx_suite_definitions/gene_info.ipynb]: I commented it out because I was not able to figure out how to perform a regex match for each value of a list using gx validation. I'll remove it.

Comment (BWMac) [reply gx_suite_definitions/gene_info.ipynb]: If you wanted to do this you would need to use the JSON schema validation expectation https://github.com/Sage-Bionetworks/agora-data-tools/pull/160#pullrequestreview-2458858218

File: tests/transform/test_gene_info.py [hunk 12]
```diff
@@ -227,6 +233,23 @@ class TestTransformGeneInfo:
         "Fail with bad data type in tep_adi_info's is_adi column",
         "Fail with bad data type in tep_adi_info's is_tep column",
     ]
+    unpiprot_input_files = {
```
Comment (jaclynbeck-sage) [inline tests/transform/test_gene_info.py]: I think this section goes away since the test for the function no longer exists?

File: tests/test_assets/gene_info/input/geneinfo_uniprot_good.csv [hunk 13]
```diff
@@ -0,0 +1,6 @@
+ensembl_gene_id,other_column1,other_column2
```
Comment (jaclynbeck-sage) [inline tests/test_assets/gene_info/input/geneinfo_uniprot_good.csv]: Is this file still needed?

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
Just cleanup of unused variables and files for testing, everything else looks great!

File: tests/test_assets/gene_info/input/geneinfo_uniprot_good.csv (hunk 13 above)
Comment (beatrizsaldana) [reply tests/test_assets/gene_info/input/geneinfo_uniprot_good.csv]: Oops no it is not!

Comment (beatrizsaldana) [reply tests/test_assets/gene_info/input/geneinfo_uniprot_good.csv:1]: Wait, yes it is!

File: tests/transform/test_gene_info.py (hunk 12 above)
Comment (jaclynbeck-sage) [reply tests/transform/test_gene_info.py]: Can you confirm if this section should go away, or stay?

**Review by BWMac — APPROVED** [review]
🚀 Pre-approving pending finishing up Jaclyn's comments

**Review by jaclynbeck-sage — APPROVED** [review]
Looks good, approved!
